File: src/get_token.py
"""Exchange an Upstox authorization code for an access token (non-interactive).

Usage:
    python src/get_token.py "<code-or-full-redirect-url>"

Reads credentials from .env and saves the token to .access_token.
This is the script for the "paste the redirect URL" workflow (auth.py is the
interactive equivalent you run in your own terminal).
"""

import os
import sys
import urllib.parse

from dotenv import load_dotenv
import upstox_client
from upstox_client.rest import ApiException

PROJECT_ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
load_dotenv(os.path.join(PROJECT_ROOT, ".env"))
KEY = os.getenv("UPSTOX_API_KEY")
SECRET = os.getenv("UPSTOX_API_SECRET")
REDIRECT = os.getenv("UPSTOX_REDIRECT_URI")
API_VERSION = "2.0"
TOKEN_FILE = os.path.join(PROJECT_ROOT, ".access_token")


def extract_code(s):
    s = s.strip().strip('"').strip("'")
    if s.startswith("http"):
        qs = urllib.parse.parse_qs(urllib.parse.urlparse(s).query)
        return (qs.get("code") or [None])[0]
    return s


def main():
    if len(sys.argv) < 2:
        print('Usage: python src/get_token.py "<code-or-redirect-url>"')
        sys.exit(1)

    code = extract_code(sys.argv[1])
    if not code:
        print("ERROR: could not extract a 'code' from the input.")
        sys.exit(1)

    try:
        resp = upstox_client.LoginApi().token(
            API_VERSION,
            code=code,
            client_id=KEY,
            client_secret=SECRET,
            redirect_uri=REDIRECT,
            grant_type="authorization_code",
        )
    except ApiException as e:
        print("TOKEN EXCHANGE FAILED")
        print(f"  Status: {e.status}  Reason: {e.reason}")
        print(f"  Body:   {e.body}")
        print("\n  UDAPI100058 => account segments inactive; reactivate in the Upstox app.")
        print("  'invalid'/credentials => check API secret in .env.")
        print("  code used/expired => log in again for a fresh code.")
        sys.exit(1)

    with open(TOKEN_FILE, "w", encoding="utf-8") as f:
        f.write(resp.access_token)

    print(f"SUCCESS: access token saved to {TOKEN_FILE}")
    print("Next: python src/market_snapshot.py   (then ask the fno-strategist agent for a trade)")


if __name__ == "__main__":
    main()

File: src/main/java/com/zaheenmunshi/upstox/GetToken.java
package com.zaheenmunshi.upstox;

import com.upstox.ApiException;
import com.upstox.api.TokenResponse;
import io.swagger.client.api.LoginApi;

import java.net.URI;
import java.net.URLDecoder;
import java.nio.charset.StandardCharsets;

/**
 * Port of {@code get_token.py}: exchange an Upstox authorization code for an
 * access token (non-interactive). Reads credentials from {@code .env} and saves
 * the token to {@code .access_token}.
 *
 * <pre>java -jar upstox-fno-pipeline.jar get-token "&lt;code-or-full-redirect-url&gt;"</pre>
 */
public final class GetToken {

    private GetToken() {}

    public static void main(String[] args) {
        if (args.length < 1) {
            System.out.println("Usage: get-token \"<code-or-redirect-url>\"");
            System.exit(1);
        }

        Config cfg = Config.load();
        String code = extractCode(args[0]);
        if (code == null || code.isBlank()) {
            System.out.println("ERROR: could not extract a 'code' from the input.");
            System.exit(1);
        }

        TokenResponse resp;
        try {
            resp = new LoginApi().token(
                    Config.API_VERSION,
                    code,
                    cfg.apiKey(),
                    cfg.apiSecret(),
                    cfg.redirectUri(),
                    "authorization_code");
        } catch (ApiException e) {
            System.out.println("TOKEN EXCHANGE FAILED");
            System.out.println("  Status: " + e.getCode());
            System.out.println("  Body:   " + e.getResponseBody());
            System.out.println("\n  UDAPI100058 => account segments inactive; reactivate in the Upstox app.");
            System.out.println("  'invalid'/credentials => check API secret in .env.");
            System.out.println("  code used/expired => log in again for a fresh code.");
            System.exit(1);
            return;
        }

        cfg.saveToken(resp.getAccessToken());
        System.out.println("SUCCESS: access token saved to " + cfg.tokenFile());
        System.out.println("Next: snapshot   (then ask the fno-strategist agent for a trade)");
    }

    /** Accept either a raw code or the full redirect URL and extract the code. */
    static String extractCode(String s) {
        s = s.strip();
        if (s.startsWith("\"") || s.startsWith("'")) s = s.replaceAll("^[\"']|[\"']$", "");
        if (s.startsWith("http")) {
            String query = URI.create(s).getQuery();
            if (query == null) return null;
            for (String pair : query.split("&")) {
                int eq = pair.indexOf('=');
                if (eq > 0 && pair.substring(0, eq).equals("code")) {
                    return URLDecoder.decode(pair.substring(eq + 1), StandardCharsets.UTF_8);
                }
            }
            return null;
        }
        return s;
    }
}
